File: lab9ZastavniyKI305/KI305/Zastavnyi/lab9/Plant.py
import os
from Leaf import Leaf
from Size import Size
from Fruit import Fruit

class Plant:
    def __init__(self):
        """
        Конструктор класу Plant.

        Ініціалізує атрибути рослини.

        Attributes:
        - name (str): Ім'я рослини.
        - size (Size): Об'єкт класу Size, представляє розмір рослини.
        - color (str): Колір рослини.
        - fruit (Fruit): Об'єкт класу Fruit, представляє фрукти рослини.
        - leaf (Leaf): Об'єкт класу Leaf, представляє листя рослини.
        - humidity (float): Вологість рослини.
        - count_of_leaves (int): Кількість листків рослини.
        - fout (File): Файловий об'єкт для ведення логу.

        """
        self.name = "Немає"
        self.size = Size()
        self.color = "Немає"
        self.fruit = Fruit()
        self.leaf = Leaf()
        self.humidity = 0
        self.count_of_leaves = 0
        self.fout = open("./Log.txt", "a")

    def __del__(self):
        """
        Деструктор класу Plant.

        Закриває відкритий файл і завершує запис у файл логу.

        """
        if hasattr(self, 'fout') and self.fout:
            self.finish()

    def finish(self):
        """
        Завершення запису у файл логу.

        Закриває файл, якщо він відкритий.

        """
        if hasattr(self, 'fout') and self.fout:
            self.fout.close()

    def grow_up(self):
        """
        Збільшення розміру рослини та кількості листя при відповідних умовах.

        Raises:
        - ValueError: Якщо вологість рослини менше або дорівнює 0.

        """
        if self.humidity <= 0:
            raise ValueError("Рослина не може рости: недостатньо вологості.")
        else:
            self.size.change_size_to_grow()
            self.set_count_of_leaves(self.count_of_leaves + 1)

    def set_count_of_leaves(self, count_of_leaves):
        """
        Встановлення кількості листя рослини.

        Args:
        - count_of_leaves (int): Кількість листя.

        Raises:
        - ValueError: Якщо передано від'ємне значення.

        """
        if count_of_leaves < 0:
            raise ValueError("Вказано неправильне значення кількості листків")
        elif count_of_leaves == 0:
            self.count_of_leaves = count_of_leaves
            self.leaf.state = True
        else:
            self.count_of_leaves = count_of_leaves

    def get_status(self):
        """
        Виведення статусу рослини.

        Виводить інформацію про ім'я, колір, вологість, розмір, кількість листя, листя та фрукти.

        """
        print("Рослина:", self.name)
        print("Колір:", self.color)
        print("Вологість:", self.humidity)

        self.fout.write("Рослина: " + self.name + "\n")
        self.fout.write("Колір: " + self.color + "\n")
        self.fout.write("Вологість: " + str(self.humidity) + "\n")

        self.size.get_status(self.fout)

        print("Кількість листків:", self.count_of_leaves)

        self.fout.write("Кількість листків: " + str(self.count_of_leaves) + "\n")

        self.leaf.get_status(self.fout)
        self.fruit.get_status(self.fout)
        print()
        self.fout.write("\n")

    def photosynthesis(self, count_light):
        """
        Виконання процесу фотосинтезу рослини.

        Args:
        - count_light (float): Кількість світла.

        Raises:
        - ValueError: Якщо передано від'ємне або нульове значення кількості світла.

        """
        if count_light <= 0:
            raise ValueError("Недостатньо світла для фотосинтезу.")
        elif self.humidity < count_light * 0.1 or self.count_of_leaves <= 0:
            raise ValueError("Недостатньо вологості або листя для фотосинтезу.")
        else:
            self.humidity -= count_light * 0.1
            print("Сфотосинтезовано", count_light * 1.2, "кисню.")

    def death(self):
        """
        Процес вмирання рослини.

        Змінює розмір рослини, колір, кількість листя та вологість.

        """
        self.size.change_size_to_death()
        self.set_color("Немає")
        self.set_count_of_leaves(0)
        self.set_humidity(0)
        self.finish()

    def set_name(self, name):
        """
        Встановлення ім'я рослини.

        Args:
        - name (str): Ім'я рослини.

        """
        self.name = name

    def set_color(self, color):
        """
        Встановлення коліру рослини.

        Args:
        - color (str): Колір рослини.

        """
        self.color = color

    def set_humidity(self, humidity):
        """
        Встановлення вологості рослини.

        Args:
        - humidity (float): Вологість рослини.

        Raises:
        - ValueError: Якщо передано від'ємне значення вологості.

        """
        if humidity < 0:
            raise ValueError("Вологість не може бути від'ємною.")
        self.humidity = humidity

    def set_fruit(self, fruit):
        """
        Встановлення об'єкта фруктів для рослини.

        Args:
        - fruit (Fruit): Об'єкт класу Fruit.

        """
        self.fruit = fruit

    def set_leaf(self, leaf):
        """
        Встановлення об'єкта листя для рослини.

        Args:
        - leaf (Leaf): Об'єкт класу Leaf.

        """
        self.leaf = leaf

    def set_size(self, size):
        """
        Встановлення об'єкта розміру для рослини.

        Args:
        - size (Size): Об'єкт класу Size.

        """
        self.size = size

    def set_count_of_leaves(self, count_of_leaves):
        """
        Встановлення кількості листя для рослини.

        Args:
        - count_of_leaves (int): Кількість листя.

        """
        self.count_of_leaves = count_of_leaves

    def set_fout(self, fout):
        """
        Встановлення файлового об'єкта для запису у файл логу.

        Args:
        - fout (File): Файловий об'єкт.

        """
        self.fout = fout


# # Приклад використання:
# try:
#     plant1 = Plant()
#     plant1.set_name("Соняшник")
#     plant1.set_color("Жовтий")
#     plant1.set_count_of_leaves(10)
#     plant1.set_humidity(50.0)
#     plant1.set_leaf(Leaf("n","true","m"))
#     plant1.grow_up()
#     plant1.get_status()
#     plant1.photosynthesis(5.0)
#     plant1.death()
#
# except Exception as e:
#     print("Помилка:", e)

File: Lab3ZastavnyiKI305/Lab3ZastavnyiKI305/src/KI305/Zastavnyi/Lab3/Plant.java
/**
 * KI305.Zastavnyi.Lab2package
 */
package KI305.Zastavnyi.Lab3;
import java.io.*;
import static java.lang.System.out;
/**
 * Клас, що представляє абстрактний об'єкт "Рослина". Реалізує функціонал рослини та містить абстрактні методи,
         * які повинні бути реалізовані в конкретних рослинах.
         *
         * @author Roman Zastavnyi
        * @version 1.0
        */
public abstract class Plant {
    private String name;            // Ім'я рослини
    private Size size;              // Розмір рослини
    private String color;           // Колір рослини
    private Fruit fruit;            // Плід рослини
    private Leaf leaf;              // Листок рослини
    private double humidity;        // Вологість рослини
    private int countOfLeaves;      // Кількість листків
    private PrintWriter fout;       // Об'єкт для запису у файл

    /**
     * Конструктор за замовчуванням. Ініціалізує всі властивості рослини на значення за замовчуванням та створює
     * об'єкт PrintWriter для запису у файл "Log.txt".
     *
     * @throws IOException
     */
    public Plant() throws IOException {
        this.name = "Немає";
        this.size = new Size();
        this.color = "Немає";
        this.fruit = new Fruit();
        this.leaf = new Leaf();
        this.humidity = 0;
        this.countOfLeaves = 0;
        fout = new PrintWriter(new FileWriter("Log.txt", true));
    }

    /**
     * Конструктор, що дозволяє встановити ім'я, колір та кількість листків рослини при створенні об'єкта.
     *
     * @param _name          Ім'я рослини
     * @param _color         Колір рослини
     * @param countOfLeaves  Кількість листків
     * @throws IOException
     */
    public Plant(String _name, String _color, int countOfLeaves) throws IOException {
        this.name = _name;
        this.size = new Size();
        this.color = _color;
        this.fruit = new Fruit();
        this.leaf = new Leaf(true);
        this.humidity = 0;
        this.countOfLeaves = countOfLeaves;
        fout = new PrintWriter(new FileWriter("Log.txt", true));
    }

    /**
     * Конструктор, що дозволяє встановити всі властивості рослини при створенні об'єкта.
     *
     * @param name           Ім'я рослини
     * @param size           Розмір рослини
     * @param color          Колір рослини
     * @param fruit          Плід рослини
     * @param leaf           Листок рослини
     * @param humidity       Вологість рослини
     * @param countOfLeaves  Кількість листків
     * @throws IOException
     */
    public Plant(String name, Size size, String color, Fruit fruit, Leaf leaf, double humidity, int countOfLeaves) throws IOException {
        this.name = name;
        this.size = size;
        this.color = color;
        this.fruit = fruit;
        this.leaf = leaf;
        this.humidity = humidity;
        this.countOfLeaves = countOfLeaves;
        fout = new PrintWriter(new FileWriter("Log.txt", true));
    }

    /**
     * Mетод, який повертає PrintWriter об'єкт для запису у файл.
     *
     * @return PrintWriter об'єкт
     */
    public PrintWriter getFout()
    {
        return fout;
    }

    /**
     * Метод, що повертає розмір рослини.
     *
     * @return Розмір рослини
     */
    public Size getSize() {
        return size;
    }

    /**
     * Метод, що встановлює розмір рослини.
     *
     * @param size Розмір рослини
     */
    public void setSize(Size size) {
        this.size = size;
    }

    /**
     * Метод, що повертає плід рослини.
     *
     * @return Плід рослини
     */
    public Fruit getFruit() {
        return fruit;
    }

    /**
     * Метод, що встановлює плід рослини.
     *
     * @param fruit Плід рослини
     */
    public void setFruit(Fruit fruit) {
        this.fruit = fruit;
    }

    /**
     * Метод, що повертає листок рослини.
     *
     * @return Листок рослини
     */
    public Leaf getLeaf() {
        return leaf;
    }

    /**
     * Метод, що встановлює листок рослини.
     *
     * @param leaf Листок рослини
     */
    public void setLeaf(Leaf leaf) {
        this.leaf = leaf;
    }

    /**
     * Метод, що встановлює ім'я рослини.
     *
     * @param name Ім'я рослини
     */
    public void setName(String name) {
        this.name = name;
    }

    /**
     * Метод, що встановлює колір рослини.
     *
     * @param color Колір рослини
     */
    public void setColor(String color) {
        this.color = color;
    }

    /**
     * Метод, що встановлює вологість рослини.
     *
     * @param humidity Вологість рослини
     */
    public void setHumidity(double humidity) {
        this.humidity = humidity;
    }

    /**
     * Метод, що встановлює кількість листків рослини. Якщо передано від'ємне значення, виводить повідомлення про помилку.
     * Якщо передано значення 0, встановлює кількість листків на 0 та встановлює стан листка на true.
     *
     * @param countOfLeaves Кількість листків
     */
    public void setCountOfLeaves(int countOfLeaves) {
        if (countOfLeaves > 0)
            this.countOfLeaves = countOfLeaves;
        else if (countOfLeaves == 0) {
            this.countOfLeaves = countOfLeaves;
            this.leaf.setState(true);
        } else {
            out.println("Вказано неправильне значення кількості листків");
        }
    }

    /**
     * Метод, що повертає ім'я рослини.
     *
     * @return Ім'я рослини
     */
    public String getName() {
        return name;
    }

    /**
     * Метод, що повертає колір рослини.
     *
     * @return Колір рослини
     */
    public String getColor() {
        return color;
    }

    /**
     * Метод, що повертає вологість рослини.
     *
     * @return Вологість рослини
     */
    public double getHumidity() {
        return humidity;
    }

    /**
     * Метод, що повертає кількість листків рослини.
     *
     * @return Кількість листків
     */
    public int getCountOfLeaves() {
        return countOfLeaves;
    }

    /**
     * Абстрактний метод, що представляє процес зростання рослини.
     */
    public void GrowUp() {
        if(this.humidity > 0) {
            size.changeSizeToGrow();
            setCountOfLeaves(this.countOfLeaves + 1);
        }
        else {
            out.println("Рослина не може рости: недостатньо вологості.");
        }
    }
    /**
     * Метод, що виводить статус рослини на консоль та у файл за допомогою PrintWriter.
     */
    public void getStatus() {
        out.println("Рослина: " + this.name);
        out.println("Колір: " + this.color);
        out.println("Вологість: " + this.humidity);

        fout.println("Рослина: " + this.name);
        fout.println("Колір: " + this.color);
        fout.println("Вологість: " + this.humidity);

        size.getStatus(fout);

        out.println("Кількість листків: " + this.countOfLeaves);

        fout.println("Кількість листків: " + this.countOfLeaves);

        leaf.getStatus(fout);
        fruit.getStatus(fout);
        out.println();
        fout.append("\n");
    }

    /**
     * Абстрактний метод, що представляє процес фотосинтезу рослини.
     *
     * @param countLight Кількість світла для фотосинтезу
     */
    public void photosynthesis(double countLight) {
        if (countLight > 0 && this.humidity >= countLight * 0.1 && this.humidity > 0&&this.countOfLeaves > 0) {
            this.humidity -= countLight * 0.1;
            out.println("Сфотосинтезовано " + countLight * 1.2 + " кисню.");
        } else if(this.humidity <= 0){
            out.println("Недостатньо вологості для фотосинтезу.");
        }else if(this.countOfLeaves <=0){
            out.println("Недостатньо листя для фотосинтезу.");
        }else {
            out.println("Недостатньо світла для фотосинтезу.");
        }
    }

    /**
     * Метод, що представляє процес "смерті" рослини, встановлюючи всі властивості рослини на відповідні значення.
     */
    public void Death() {
        size.changeSizeToDeath();
        setColor("Немає");
        setCountOfLeaves(0);
        setHumidity(0);
        finish();
    }

    /**
     * Метод, що викликається при завершенні роботи з рослиною та закриває PrintWriter.
     */
    public void finish() {
        fout.close();
    }
}
